bolt-simple-taproot.md

```diff
@@ -0,0 +1,1141 @@
+
+#### Taproot Script Path Spends
+
+A script path spend is executed when a tapscript leaf is revealed. A script
+path spends has three components: a control block, the script leaf being
+revealed, and the valid witness.
+
+A control block has the following serialization:
+```
+(output_key_y_parity | leaf_version) || internal_key || inclusion_proof
+```
+
+The first segment uses a spare bit of the `leaf_version` to encode the parity
+of the output key, which is checked during control block verification. The
+`internal_key` is just that. The `inclusion_proof` is the series of sibling
+hashes in the path from the revealed leaf all the way up to the root of the
+tree. In practice if one hashes the revealed leaf, and each 32-byte hash of the
+inclusion proof together, they'll reach the script root if the proof was valid.
+
+If only a single leaf is committed to in a tree, then the `inclusion_proof`
+will be absent.
+
+The final witness to spend a script path output is:
+```
+<witness1> ... <witnessN> <leaf_script> <control_block>
+```
+
+### MuSig2 & Associated Changes
+
+Musig2 is a multi-signature scheme based on schnorr signatures. It allows N
+parties to construct an aggregated public key, and then generate a single
+multi-signature that can only be generated with all the parties of the
+aggregated key. In practice we use this to allow the funding output of
+channel to look just like any other P2TR output.
+
+A musig2 signing flow has 4 stages:
+
+  1. First each party generates a public nonce, and exchanges it with every
+    other party.
+
+  2. Next public keys are exchanged by each party, and combined into a single
+    aggregated public key.
+
+  3. Next, each party generates a _partial_ signature and exchanges it with
+    every other party.
+
+  4. Finally, once all partial signatures are known, they can be combined into a
+    valid schnorr signture, which can be veirfied using BIP 340.
+
+Steps 1&2 may be performed out of order, or concurrently. In our case only two
+parties exist, so as soon as one party knows both partial signatures, they can
+be combined into a final signature.
+
+#### Key Aggregation
+
+We refer to the of `KeyAgg` algorithm [bip-musig2](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki)
+for key aggregation. In order to avoid sending extra ordering information, we
+always assume that both keys are sorted first with the `KeySort` algorithm
+before aggregating (`KeyAgg(KeySort(p1, p2))`).
+
+#### Nonce Generation
+
+A `musig2` secret nonce is the concatenation of two random, 32-byte integers.
+
+A `musig2` public nonce is technically the concatenation of two public keys, each representing the EC-point
+corresponding to its secret integer, thus resulting in a 66-byte value:
+
+```
+point_1 || point_2
+```
+
+For nonce generation, we refer to the `NonceGen` algorithm defined in the
+`musig2` BIP. We only specify the set of required arguments, though a signer can also opt to specify the additional
```

Comment:
Should we put "Nonce generation" recommendations somewhere, e.g committing to the `funding_outpoint || per_commitment_point` in `sec256k1_musig_nonce_gen`'s `extra_input32` ? Like per-commitment point should be unique for each channel state which is nice (unless you have `channel_reestablish`).

Reply:
Yep, I've tacked this on in the latest version.